docs/design.md

```diff
@@ -0,0 +1,156 @@
+depend on a chart that provides that CRD but does not provide prometheus. The
+CRD chart or your chart can declare an optional dependency on a chart that
+installs Prometheus. When using hypper to install your chart you will be asked
+about installing the optional dependencies. In development you can choose not
+to install the dependencies while in production you can choose to install them.
+
+The optional dependency handling will extend to all commands.
+
+Optional dependencies will be specified using Helm chart annotations.
+
+### Shared Dependencies
+
+Sometimes you want to have a shared system dependency. For example, you may want
+to have just a single Prometheus or Istio in a cluster. When you install an
+application that depends on this shared dependency you want to install it if not
+present and leverage the existing one if present.
+
+With Helm there is a separation between those who provide charts and those who
+consume them. Charts from various repositories can be installed in the same
+cluster but are not guaranteed to work together. That is left as an exercise for
+the chart authors.
+
+Hypper will make it easier to have a repository of charts that work together but
+can be installed separately.
+
+The additional metadata will be stored as Helm chart annotations.
+
+### OCI Storage
+
+OCI artifacts are a new way to store things. Helm supports storing Helm charts
+in OCI distributions as an experimental feature. Many container registries
+support storing non-images in them. The list includes Azure, AWS, Google Cloud,
+GitHub Container Registry, and Harbor. The one notable exception is Docker Hub.
+
+Kubernetes already needs to have a container registry. In air gapped, on-premise,
+and custom environments you need a registry to store your images. If that can be
+used to store charts instead of a separate system you have fewer system
+dependencies.
+
+Hypper will make OCI registries a primary element in the system.
+
+One notable disadvantage to using an OCI registry for storing charts is a lack of
+an `index.yaml` file. This file has metadata enabling some functionality that
+can't be duplicated with an OCI registry due to APIs not existing (e.g., search).
+To restore these features we are looking to put an index into an OCI registry.
+
+### Name As Chart Identifier
+
+Zypper, the package manager used for openSUSE and other Linux distributions,
+uses the package name as the identifier for it. You can change from one vendor
+to another for a package with some differences but still the same package.
+
+Helm does not make this form of assumption. It could be true. For example, you
+could install a chart from one location and then update from a fork in another
+location. Or two charts with the same name could be completely different. It
+is the responsibility of the end user to make this decision.
+
+Hypper is going to make the assumption that charts of the same name are the
+same chart from different locations.
+
```

Comment:
This sounds like a terrible idea, without at least checking their signatures. 
If two helm charts have the same name and signed by the same authority, that makes kind of sense to assume they have been properly handled. Else it's opening the door to problems.

Reply:
I agree, as long you rely only on one catalogue this makes sense, but as soon you have external, maybe third-party deps, this is going to cause issues.

Reply:
not more than if you were to upgrade a helm chart with the same name from a different location which does something completely different. "If you have different source, assume the helm charts are different" sounds fair to me.

Reply:
We need to think about this in practice and what's good enough. If we looked for a perfect solution than this is not a good idea.

When it comes to names, this is already used in practice. I'll give both a Helm and non-Helm example...

1. When Helm deprecated the stable charts repository people needed to move to a new location for their charts. They were able to add a new repository and just upgrade from the new one instead. Checking the API and features was left as an exercise to the end user rather than being automated. But, it worked. There are also many companies that will install a chart from a company and then later fork or vendor it. Upgrading from their fork instead.
2. In zypper you have the ability to install something and then change vendors later. The package you change to can have functional differences. For example, I change my desktop codec packages to those that support proprietary codecs so I can browse the web and use A/V. Same package name but different vendors shipping differences.

Let's think of this in terms of use cases...

* As an operator at a company, I have forked a chart we are using for our internal use. We are already using it. We want to tell the system to switch to the one we have.
* As someone who added a feature to a chart that's not accepted upstream, I want to share my new location for the chart and give people an easy way to switch to my version.
* As someone consuming a chart, the organization I get it from has stopped updating it and I want to get updates from a new location. This could be because there are security issues or other problems. But, if the organization that owns it repairs the issues I would like to switch back.

How would a system work that accommodates these types of use cases?